src/draive/splitters/basic.py

```diff
@@ -33,97 +44,91 @@ def split_text(
 
 def _split(
     text: str,
-    separators: tuple[str, str] | str | None = None,
-) -> tuple[str, list[str]]:
-    # prepare initial splitter - default is new paragraph
-    splitter: str
-    alt_splitter: str
-    match separators:
-        case (str(primary), str(secondary)):
-            splitter = primary
-            alt_splitter = secondary
-        case str(primary):
-            splitter = primary
-            alt_splitter = "\n"
-        case None:
-            splitter = "\n\n"
-            alt_splitter = "\n"
-
-    # try splitting using provided splitter
-    parts: list[str] = text.split(splitter)
-    # if splitting has done nothing retry using secondary splitter
-    if len(parts) == 1:
-        splitter = alt_splitter
-        parts = text.split(splitter)
+    splitters: Sequence[str],
+) -> tuple[str, list[str], list[str]]:
+    # try splitting using provided splitters
+    for index, splitter in enumerate(splitters):
+        parts: list[str] = text.split(splitter)
+        # if splitting has done nothing retry using next one
+        if len(parts) == 1:
+            continue
+        else:
+            # used_splitter, fallback_splitters, parts
+            return (splitter, list(splitters[index + 1 :]), parts)
     # if splitting has still done nothing then fail
-    if len(parts) == 1:
-        raise ValueError("Failed to properly split text with provided separators")
-
-    return (splitter, parts)
+    raise ValueError("Failed to properly split text with provided separators")
 
 
-def _merge(  # noqa: C901
+def _merge(  # noqa: C901, PLR0912, PLR0913
     parts: list[str],
     part_size: int,
     count_size: Callable[[str], int],
     splitter: str,
+    fallback_splitters: list[str],
     part_overlap_size: int | None,
 ) -> list[str]:
     result: list[str] = []
-    accumulator: list[str] = []
-    accumulator_size: int = 0
+    accumulator: str = ""
+    overlap_accumulator: list[str] = []
     # iterate over splitted pats
-    for part in parts:
-        temp_size: int = count_size(part)
+    last_part_index: int = len(parts) - 1
+    for index, part in enumerate(parts):
+        current_part: str
+        # Add the separator back if it's not the last part
+        if index < last_part_index:
+            current_part = part + splitter
+        else:
+            current_part = part
+
+        temp_size: int = count_size(accumulator + current_part)
+        current_part_size: int = count_size(current_part)
         # check if can add to previous part
-        if accumulator_size + temp_size < part_size:
-            accumulator.append(part)
-            accumulator_size += temp_size
+        if temp_size <= part_size:
+            accumulator += current_part
```

Comment:
we could reuse the value used to count temp_size instead of joining strings again